File: app/src/main/java/bogomolov/aa/fitrack/repository/ModelEntityMapper.java
package bogomolov.aa.fitrack.repository;

import java.util.ArrayList;
import java.util.List;

import bogomolov.aa.fitrack.core.model.Point;
import bogomolov.aa.fitrack.core.model.Tag;
import bogomolov.aa.fitrack.core.model.Track;
import bogomolov.aa.fitrack.repository.entities.PointEntity;
import bogomolov.aa.fitrack.repository.entities.TagEntity;
import bogomolov.aa.fitrack.repository.entities.TrackEntity;

public class ModelEntityMapper {

    public static Point entityToModel(PointEntity from) {
        if(from == null) return null;
        Point to = new Point();
        to.setId(from.getId());
        to.setSmoothed(from.getSmoothed());
        to.setTime(from.getTime());
        to.setLat(from.getLat());
        to.setLng(from.getLng());
        return to;
    }

    public static PointEntity modelToEntity(Point from) {
        if(from == null) return null;
        PointEntity to = new PointEntity();
        to.setId(from.getId());
        to.setSmoothed(from.getSmoothed());
        to.setTime(from.getTime());
        to.setLat(from.getLat());
        to.setLng(from.getLng());
        return to;
    }

    public static Tag entityToModel(TagEntity from) {
        if(from == null) return null;
        Tag to = new Tag();
        to.setId(from.getId());
        to.setName(from.getName());
        return to;
    }

    public static TagEntity modelToEntity(Tag from) {
        if(from == null) return null;
        TagEntity to = new TagEntity();
        to.setId(from.getId());
        to.setName(from.getName());
        return to;
    }

    public static Track entityToModel(TrackEntity from) {
        if(from == null) return null;
        Track to = new Track();
        to.setId(from.getId());
        to.setDistance(from.getDistance());
        to.setEndPoint(from.getEndPoint());
        to.setEndSmoothedPoint(from.getEndSmoothedPoint());
        to.setEndTime(from.getEndTime());
        to.setStartPoint(from.getStartPoint());
        to.setStartSmoothedPoint(from.getStartSmoothedPoint());
        to.setStartTime(from.getStartTime());
        to.setTag(from.getTag());
        return to;
    }

    public static TrackEntity modelToEntity(Track from) {
        if(from == null) return null;
        TrackEntity to = new TrackEntity();
        to.setId(from.getId());
        to.setDistance(from.getDistance());
        to.setEndPoint(from.getEndPoint());
        to.setEndSmoothedPoint(from.getEndSmoothedPoint());
        to.setEndTime(from.getEndTime());
        to.setStartPoint(from.getStartPoint());
        to.setStartSmoothedPoint(from.getStartSmoothedPoint());
        to.setStartTime(from.getStartTime());
        to.setTag(from.getTag());
        return to;
    }

    public static <Q> List<Q> entityToModel(List fromList, Class<Q> toClass) {
        List toList = new ArrayList<>();
        if (toClass.isAssignableFrom(Point.class))
            for (Object fromEntity : fromList) toList.add(entityToModel((PointEntity) fromEntity));
        if (toClass.isAssignableFrom(Tag.class))
            for (Object fromEntity : fromList) toList.add(entityToModel((TagEntity) fromEntity));
        if (toClass.isAssignableFrom(Track.class))
            for (Object fromEntity : fromList) toList.add(entityToModel((TrackEntity) fromEntity));
        return toList;
    }

    public static <Q> List<Q> modelToEntity(List fromList, Class<Q> toClass) {
        List toList = new ArrayList<>();
        if (toClass.isAssignableFrom(PointEntity.class))
            for (Object fromEntity : fromList) toList.add(modelToEntity((Point) fromEntity));
        if (toClass.isAssignableFrom(TagEntity.class))
            for (Object fromEntity : fromList) toList.add(modelToEntity((Tag) fromEntity));
        if (toClass.isAssignableFrom(TrackEntity.class))
            for (Object fromEntity : fromList) toList.add(modelToEntity((Track) fromEntity));
        return toList;
    }

}

File: app/src/main/java/bogomolov/aa/fitrack/repository/ModelEntityMapper.kt
package bogomolov.aa.fitrack.repository

import bogomolov.aa.fitrack.domain.model.Point
import bogomolov.aa.fitrack.domain.model.Tag
import bogomolov.aa.fitrack.domain.model.Track
import bogomolov.aa.fitrack.repository.entities.PointEntity
import bogomolov.aa.fitrack.repository.entities.TagEntity
import bogomolov.aa.fitrack.repository.entities.TrackEntity
import java.util.*

fun entityToModel(from: PointEntity?): Point? {
    if (from == null) return null
    return Point(
        from.id,
        from.time,
        from.lat,
        from.lng,
        from.smoothed
    )
}

fun modelToEntity(from: Point) =
    PointEntity(
        from.id,
        from.time,
        from.lat,
        from.lng,
        from.smoothed
    )

fun entityToModel(from: TagEntity?): Tag? {
    if (from == null) return null
    return Tag(from.id, from.name ?: "")
}

fun modelToEntity(from: Tag) = TagEntity(from.id, from.name)

fun entityToModel(from: TrackEntity?): Track? {
    if (from == null) return null
    return Track(
        from.id,
        from.startPointId,
        from.endPointId,
        from.startSmoothedPointId,
        from.endSmoothedPointId,
        from.startTime,
        from.endTime,
        from.tag,
        from.distance
    )
}

fun modelToEntity(from: Track) =
    TrackEntity(
        from.id,
        from.startPointId,
        from.endPointId,
        from.startSmoothedPointId,
        from.endSmoothedPointId,
        from.startTime,
        from.endTime,
        from.tag,
        from.distance
    )


fun <Q> entityToModel(fromList: List<*>, toClass: Class<Q>): List<Q> {
    val toList: MutableList<Q> = ArrayList()
    if (toClass.isAssignableFrom(Point::class.java))
        for (fromEntity in fromList) toList.add(entityToModel(fromEntity as PointEntity) as Q)
    if (toClass.isAssignableFrom(Tag::class.java))
        for (fromEntity in fromList) toList.add(entityToModel(fromEntity as TagEntity) as Q)
    if (toClass.isAssignableFrom(Track::class.java))
        for (fromEntity in fromList) toList.add(entityToModel(fromEntity as TrackEntity) as Q)
    return toList
}